### ISSUE TITLE
How to enable U1000 (unused) for staticcheck

### ISSUE THREAD

**Issue by Sean-Der** [sluongng/nogo-analyzer#46]
If I make the following change to the pass example the build doesn't fail

```
diff --git a/examples/staticcheck/cmd/pass/main.go b/examples/staticcheck/cmd/pass/main.go
index ffb6147..265d37d 100644
--- a/examples/staticcheck/cmd/pass/main.go
+++ b/examples/staticcheck/cmd/pass/main.go
@@ -5,6 +5,11 @@ import (
        "time"
 )
 
+const x = "foo"
+
+func deadcode() {
+}
+
 func main() {
        d, err := time.Parse("01-02-2006", "01-01-2023")
        if err != nil {
```

`U1000` is enabled in `ANALYZERS` in `def.bzl` though. I don't understand yet what's happening, going to debug further.

**Comment by sluongng** [comment]
That's a good report.

I digged a bit into it and it's because the `unused` Analyzer in staticcheck does not report the findings through `pass.Report`

https://github.com/dominikh/go-tools/blob/a2bad015c062e0d4ffabc65eb948e76acc395f0d/unused/unused.go#L209-L230

I fixed the code to this
```
func run(pass *analysis.Pass) (interface{}, error) {
	g := newGraph(
		pass.Fset,
		pass.Files,
		pass.Pkg,
		pass.TypesInfo,
		pass.ResultOf[directives.Analyzer].([]lint.Directive),
		pass.ResultOf[generated.Analyzer].(map[string]generated.Generator),
		DefaultOptions,
	)
	g.entry()

	sg := &SerializedGraph{
		nodes: g.nodes,
	}

	if Debug != nil {
		Debug.Write([]byte(sg.Dot()))
	}

	result := sg.Results()
	for _, u := range result.Unused {
		pass.Report(analysis.Diagnostic{
			Pos:      u.Pos, // <- NEW FIELD
			Category: "U1000",
			Message:  "unused: " + u.Name,
		})
	}

	return result, nil
}
...
func (g *graph) objectToObject(obj types.Object) Object {
	...
	return Object{
		Name:            name,
		ShortName:       obj.Name(),
		Kind:            typString(obj),
		Path:            path,
		Pos:             obj.Pos(), // <- NEW FIELD
		Position:        g.fset.PositionFor(obj.Pos(), false),
		DisplayPosition: report.DisplayPosition(g.fset, obj.Pos()),
	}
}
...
type Object struct {
	Name      string
	ShortName string
	// OPT(dh): use an enum for the kind
	Kind            string
	Path            ObjectPath
	Pos             token.Pos // <- NEW FIELD
	Position        token.Position
	DisplayPosition token.Position
}
```

which simply loops over `sg.Results().Unused` and call `pass.Report` for each findings and it should work correctly. Here is a quick run in BuildBuddy codebase.

```
ERROR: /Users/sluongng/work/buildbuddy/buildbuddy/server/util/blocklist/BUILD:3:11: Validating nogo output for //server/util/blocklist:blocklist failed: (Exit 1): builder failed: error executing ValidateNogo command (from target //server/util/blocklist:blocklist)
  (cd /private/var/tmp/_bazel_sluongng/06e573a93bc2d6a9cad4ad41f00b4310/execroot/buildbuddy && \
  exec env - \
  bazel-out/platform_linux_x86_64-opt-exec/bin/external/go_sdk_linux/builder_reset/builder nogovalidation bazel-out/platform_linux_x86_64-fastbuild/bin/server/util/blocklist/blocklist.nogo bazel-out/platform_linux_x86_64-fastbuild/bin/server/util/blocklist/blocklist.nogo.log bazel-out/platform_linux_x86_64-fastbuild/bin/server/util/blocklist/blocklist.nogo.patch)
# Configuration: d5a1f172902c0252ebe818ee8412a7051d17da7c53d8fd7c40982972040364bd
# Execution platform: @@buildbuddy_toolchain//:platform_linux_x86_64

nogo: errors found by nogo during build-time code analysis:
server/util/blocklist/blocklist.go:8:6: unused: hashGroupID (U1000)
saving suggested fixes:
some suggested fixes are invalid or have conflicts with other fixes:
	ignoring suggested fixes from analyzer "U1000" at server/util/blocklist/blocklist.go:8:6
please apply other fixes and rerun the build.
```

I am currently out of bandwidth to drive this change upstream, though. Feel free to report it upstream so other folks might come up with a better fix.

**Comment by sluongng** [comment]
hmm a better message would be
```
			Message:  fmt.Sprintf("unused %s: %q", u.Kind, u.Name),
```

which yields

```
nogo: errors found by nogo during build-time code analysis:
enterprise/server/usage/usage_test.go:40:2: unused var: "period4Start" (U1000)
enterprise/server/usage/usage_test.go:665:6: unused func: "addCounts" (U1000)
saving suggested fixes:
some suggested fixes are invalid or have conflicts with other fixes:
	ignoring suggested fixes from analyzer "U1000" at enterprise/server/usage/usage_test.go:40:2
	ignoring suggested fixes from analyzer "U1000" at enterprise/server/usage/usage_test.go:665:6
please apply other fixes and rerun the build.
```

**Comment by sluongng** [comment]
I created a short and sweet issue upstream here https://github.com/dominikh/go-tools/issues/1640

Feel free to play around and submit a PR there.

**Comment by Sean-Der** [comment]
@sluongng Would you be open to a PR that replicated the logic needed for unused here?

It shouldn't be that much

**Comment by sluongng** [comment]
As tempting as it is, I don't have the capacity to maintain a fork of the analyzer right now.

I think it would be best if you create a thin Analyzer wrapper on top of https://github.com/dominikh/go-tools/blob/master/unused/unused.go and loop through the `Result.Unused` objects. Feel free to share the result here so others can follow and use your wrapper.

**Comment by joneshf-dd** [comment]
Heya! Ran into something similar while trying to figure out how `unused` is supposed to work. We've had a wrapper that looks similar to this for a while:
```Go
import (
	"fmt"
	"go/token"

	"golang.org/x/tools/go/analysis"
	"honnef.co/go/tools/unused"
)

var Analyzer = &analysis.Analyzer{
	Name:       unused.Analyzer.Analyzer.Name,
	Doc:        unused.Analyzer.Analyzer.Doc,
	Requires:   unused.Analyzer.Analyzer.Requires,
	ResultType: unused.Analyzer.Analyzer.ResultType,
	Run: func(pass *analysis.Pass) (interface{}, error) {
		out, err := unused.Analyzer.Analyzer.Run(pass)
		if err != nil {
			return nil, err
		}

		results, ok := out.(unused.Result)
		if !ok {
			return nil, fmt.Errorf("unexpected result type: %T", out)
		}

		for _, obj := range results.Unused {
			pass.Fset.Iterate(func(file *token.File) bool {
				if file.Name() != obj.Position.Filename {
					return true
				}

				pass.Report(analysis.Diagnostic{
					Message: fmt.Sprintf("%s %s is unused", obj.Kind, obj.Name),
					Pos:     file.Pos(obj.Position.Offset),
				})
				return false
			})
		}

		return out, nil
	},
}
```

But as mentioned in the "unexported things being used in other code"-thing explained here: https://github.com/dominikh/go-tools/issues/1640#issuecomment-2722041760, it doesn't provide accurate results. Because of the size of our code base, we ended up having to disable it basically everywhere to draw a line in the sand. Our thought at the time was that this pattern isn't ideal in the first place: unexported things should either be exported and treated like a public API or not depended on at all (even in internal tests). It's an opinion, not a law; but what we went with at the time.

Thinking about how to make this check useful, the suggestion is to do analysis on more than a single package. In the `nogo` world, this means analysis on more than a single `go_binary`/`go_library`/etc. target. It seems like this is possible in some cases, but not all. It would seem that based on the type of target, and its visibility you can determine first whether you can even validate this correctly. If you can validate it, you should be able to determine which packages you need to validate it. The main thing that seems to influence whether you can validate something is its potential reverse dependencies. In the absence of being able to limit reverse dependencies, you cannot tell if something is truly unused.

- If a `.go` file is used in an `exports_files`/`filegroup`/`go_library`/`go_source` or similar target that either provides a `DefaultInfo` provider with the `.go` file, `GoInfo` provider, or some other provider that allows access to the actual `.go` file; there are two cases to look at:
    - If the target has public `visibility`, you cannot validate unused things.
        - Reason being, that `.go` file can be used anywhere else someone has access to it, including a different Bazel repository. In that other place, they may be using the thing that looks unused. This could be done by using it directly as `srcs`, or with like `embed` if it meets the restrictions.
    - If the target does not have a public `visibility`, you can gather all of the reverse dependencies and validate that there are no unused things between all of them.
- If a `.go` file  is only used in a `go_binary`/`go_test` or similar target that cannot be embedded in another target in a way that would re-compile the source code, then `unused` can be used directly to validate that there are no unused things.

At least, that's what I'm thinking would be the proper way to use `unused` with `nogo`. We haven't actually implemented any of this to see where it falls down (I'm sure it's missing some edge cases). If anyone comes up with a solution, we'd be interested in how it goes!
